src/enzyme_ad/jax/Implementations/StableHLOAutoDiffOpInterfaceImpl.cpp

```diff
@@ -616,14 +616,282 @@ class AutoDiffConcatenateRev
+      .getResult();
+}
+
+static mlir::TensorType applyBatchSizes(mlir::Type Ty,
+                                        llvm::ArrayRef<int64_t> batchSizes) {
+  auto T = cast<TensorType>(Ty);
+  SmallVector<int64_t> shape(batchSizes.begin(), batchSizes.end());
+  shape.append(T.getShape().begin(), T.getShape().end());
+  auto T2 = T.clone(shape);
+  return T2;
+}
+
+// TODO: make public in Enzyme MLIR?
+// this is essentially
+// https://github.com/EnzymeAD/Enzyme/blob/342057e3a3e657a33da8295c99acdcd20b0375f4/enzyme/Enzyme/MLIR/Passes/EnzymeBatchPass.cpp#L58-L100
+static void batchCloneBlock(Block *srcBlock, Block *destBlock,
+                            IRMapping &mapper, ArrayRef<int64_t> batchSizes) {
+  for (auto arg : srcBlock->getArguments()) {
+    auto batched = destBlock->addArgument(
+        applyBatchSizes(arg.getType(), batchSizes), arg.getLoc());
+    mapper.map(arg, batched);
+  }
+
+  bool valid = true;
+
+  OpBuilder builder(destBlock, destBlock->end());
+  for (auto &src : srcBlock->getOperations()) {
+    if (auto ifaceOp = dyn_cast<BatchOpInterface>(&src)) {
+      auto res = ifaceOp.createBatch(builder, mapper, batchSizes);
+      if (res.succeeded())
+        continue;
+    }
+
+    SmallVector<Value, 8> operands;
+    SmallVector<Block *, 2> successors;
+
+    // Remap the operands.
+    operands.reserve(src.getNumOperands());
+    for (auto opValue : src.getOperands())
+      operands.push_back(mapper.lookup(opValue));
+
+    // Remap the successors.
+    successors.reserve(src.getNumSuccessors());
+    for (Block *successor : src.getSuccessors())
+      successors.push_back(mapper.lookup(successor));
+
+    SmallVector<Type> resultTypes(src.getResultTypes().begin(),
+                                  src.getResultTypes().end());
+    for (auto &Ty : resultTypes) {
+      Ty = applyBatchSizes(Ty, batchSizes);
+    }
+
+    Operation *newOp = Operation::create(
+        src.getLoc(), src.getName(), resultTypes, operands, src.getAttrs(),
+        OpaqueProperties(nullptr), successors, src.getNumRegions());
+
+    // // Clone the regions.
+    // for (auto &&[oldReg, newReg] :
+    //      llvm::zip(src.getRegions(), newOp->getRegions())) {
+    //   batchCloneRegion(&oldReg, &newReg, mapper, batchSizes);
+    // }
+
+    // Remember the mapping of any results.
+    for (unsigned i = 0, e = src.getNumResults(); i != e; ++i)
+      mapper.map(src.getResult(i), newOp->getResult(i));
+
+    builder.insert(newOp);
+  }
+}
+
+// For some ops with nested regions, identify if we can batch the inner regions
+// instead
+static LogicalResult tryToBatchInner(Operation *src, OpBuilder &builder,
+                                     IRMapping &mapper,
+                                     ArrayRef<int64_t> batchSizes) {
+  if (auto ifOp = dyn_cast<IfOp>(src)) {
+    auto iszero = matchPattern(ifOp.getPred(), m_Zero());
+    auto isone = matchPattern(ifOp.getPred(), m_One());
+
+    if (!iszero && !isone)
```

Comment:
can we also test if the input operand would be a broadcastop of a single value?

Reply:
added it but I am having a hard time coming up with an input to test this where the batched version of the predicate is a broadcast_in_dim.

Reply:
yeah I think long term this will require an extension to the batching interface to mark some operands as non-batched, in any case is fine for now